File: problems/948/solution.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int P) {
        int score = 0;
        int lo = 0;
        int hi = tokens.length - 1;
        
        Arrays.sort(tokens);
        
        while (lo <= hi) {
            if (tokens[lo] <= P) {
                P -= tokens[lo];
                score++;
                lo++;
            } else if (score > 0) {
                if (lo == hi) {
                    break;
                }
                
                P += tokens[hi];
                score--;
                hi--;
            } else {
                break;
            }
        }
        
        return score;
    }
}

File: problems/948/solution.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], P: int) -> int:
        score, lo, hi = 0, 0, len(tokens) - 1
        
        tokens.sort()
        
        while lo <= hi:
            if tokens[lo] <= P:
                P -= tokens[lo]
                score += 1
                lo += 1
            elif score > 0:
                if lo == hi:
                    break
                
                P += tokens[hi]
                score -= 1
                hi -= 1
            else:
                break
            
        return score